### PR TITLE
util.container.array: Add invariant for _ptr/_length

### PR DESCRIPTION
**Pull request by klickverbot**
This was quite useful for debugging a memory corruption issue.

### CODE REVIEW

**Comment by ZombineDev** [comment]
LGTM

File: src/rt/util/container/array.d [hunk 1]
```diff
@@ -39,11 +39,11 @@ nothrow:
...
-            if (nlength < length)
-                foreach (ref val; _ptr[nlength .. length]) common.destroy(val);
+            if (nlength < _length)
+                foreach (ref val; _ptr[nlength .. _length]) common.destroy(val);
             _ptr = cast(T*)common.xrealloc(_ptr, reqsize);
-            if (nlength > length)
-                foreach (ref val; _ptr[length .. nlength]) common.initialize(val);
+            if (nlength > _length)
+                foreach (ref val; _ptr[_length .. nlength]) common.initialize(val);
             _length = nlength;
```
Comment (schveiguy) [inline src/rt/util/container/array.d]: Wouldn't the invariant fail here if `_length` is set to 0, but `_ptr` is not set to null?

Comment (schveiguy) [inline src/rt/util/container/array.d]: oops, nevermind. I see _ptr is realloc above.

**Comment by schveiguy** [comment]
Yes, I looked that up too, was going to post about it but realized it's good :)

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Pull request merged**